File: src/commands/sync.ts
/**
 * Sync command.
 */
import { Command } from "commander";
import { getClient } from "../api/client.js";
import { printInfo, printJson, printKeyValue } from "../output/index.js";
import { handleError } from "./errors.js";
import { getGlobalOptions } from "../index.js";

export function createSyncCommand(): Command {
  const sync = new Command("sync").description("Sync and backup");

  // sync command (default)
  sync
    .command("all", { isDefault: true })
    .description("Get full state snapshot (tasks, projects, tags, groups)")
    .option("--json", "Output raw JSON (useful for backups)")
    .action(
      handleError(async function (this: Command, options) {
        const globalOpts = getGlobalOptions(this);
        const client = await getClient({ validation: globalOpts.validation });
        const batch = await client.getBatch();

        if (options.json) {
          printJson(batch);
        } else {
          const tasks = batch.syncTaskBean?.update ?? [];
          const projects = batch.projectProfiles ?? [];
          const tags = batch.tags ?? [];
          const groups = batch.projectGroups ?? [];

          printKeyValue(
            {
              "Active Tasks": tasks.length.toString(),
              "Projects": projects.length.toString(),
              "Tags": tags.length.toString(),
              "Project Groups": groups.length.toString(),
              "Inbox ID": batch.inboxId ?? "-",
              "Checkpoint": batch.checkPoint?.toString() ?? "-",
            },
            [
              "Active Tasks",
              "Projects",
              "Tags",
              "Project Groups",
              "Inbox ID",
              "Checkpoint",
            ]
          );

          printInfo("\nUse --json flag to export full data for backup.");
        }
      })
    );

  return sync;
}

File: ticktick_cli/cli/sync.py
"""Sync command for TickTick CLI."""

from typing import Annotated

import typer

from ticktick_cli import output
from ticktick_cli.cli.errors import handle_client_error
from ticktick_cli.client import get_client

app = typer.Typer(help="Sync operations")


@app.callback(invoke_without_command=True)
@handle_client_error
def sync(
    ctx: typer.Context,
    json_output: Annotated[
        bool,
        typer.Option("--json", help="Output as JSON"),
    ] = False,
) -> None:
    """Fetch full state snapshot from TickTick.

    Returns all tasks, projects, tags, and other data in a single request.
    Useful for syncing or backup purposes.
    """
    if ctx.invoked_subcommand is not None:
        return

    client = get_client()
    data = client.get_batch()

    if json_output:
        output.print_json(data)
    else:
        # Provide a summary instead of dumping everything
        tasks = data.get("syncTaskBean", {}).get("update", []) or []
        projects = data.get("projectProfiles", []) or []
        tags = data.get("tags", []) or []
        groups = data.get("projectGroups", []) or []

        output.print_info("Sync complete. Summary:")
        output.print_key_value(
            {
                "Tasks": str(len(tasks)),
                "Projects": str(len(projects)),
                "Tags": str(len(tags)),
                "Project Groups": str(len(groups)),
            },
            ["Tasks", "Projects", "Tags", "Project Groups"],
        )
        output.print_info("\nUse --json for full data export.")
